library/ssl_tls.c

```diff
@@ -622,6 +622,55 @@ static int tls1_prf( const unsigned char *secret, size_t slen,
 
 #if defined(MBEDTLS_SSL_PROTO_TLS1_2)
 #if defined(MBEDTLS_USE_PSA_CRYPTO)
+
+static psa_status_t setup_psa_key_derivation( psa_key_derivation_operation_t* derivation,
+                                              psa_key_handle_t slot,
+                                              psa_algorithm_t alg,
+                                              const unsigned char* seed, size_t seed_length,
+                                              const unsigned char* label, size_t label_length,
+                                              size_t capacity )
+{
+    psa_status_t status;
+
+    status = psa_key_derivation_setup( derivation, alg );
+    if (status != PSA_SUCCESS )
+        return( status );
+
+    if( PSA_ALG_IS_TLS12_PRF( alg ) || PSA_ALG_IS_TLS12_PSK_TO_MS( alg ) )
```

Comment:
Why do we need to check this? Why can't we just rely on PSA to return an error and propagate it if something goes wrong?

Reply:
This check enables proper use of the PSA API. Please compare it with the original implementation here: https://github.com/ARMmbed/mbed-crypto/blob/development/tests/suites/test_suite_psa_crypto.function#L533

Note, that if a different algorithm is selected, we feed the API with different annotations, e.g. `PSA_KEY_DERIVATION_INPUT_SALT` vs. `PSA_KEY_DERIVATION_INPUT_SEED`.

Here, the test could be removed, but should the other algorithms be needed, all you have to do is add another branch to this `if` statement.

I can make the code executed unconditionally if you think it would be better that way.

Reply:
I like the `if` present. If we want to add HKDF, for example, the code is structured already to work with that. If we use HKDF, we need to call a different set of key derivation steps, providing different inputs at different times.